### DISCUSSION TITLE
How to configure proper math SymPy output?

### DISCUSSION THREAD

**Discussion by quctex** [quarto-dev/quarto-cli#2265] [Q&A]
Hello! 

I am trying to render Jupyter Lab notebook with Quatro to html. I am newbie both in Jupyter and Quatro (and in python too). This all is my first steps in modern programming for computations.

When I using SymPy Python library for symbolic mathematics, math output in Jupyter Lab itself and in html file, generated by Jupyter `File->Save and Export Notebook As->HTML` menu item both contains beautiful vector math formulas. But when I generate html by Quarto, formulas inserted as low-resolution png images or even ASCII-art.

**My question: how can i achieve beautiful vector formulas in Quarto-generated html file, as I seen in Jupyter Lab notebook or in html file, exported by it?**

**I found similar problem with some kind of solution [here](https://github.com/quarto-dev/quarto-cli/issues/141), but my skills so low, that I can't follow the recipe. I will be glad if someone can clarify, what changes should i make in my (minimal) file (that I will provide below) to follow that recipe. Also I see, that those solution makes python code in notebook hard to write and read, and hoping to find more transparent and direct solution.** 

To clarify, I decided use Quarto to supply my final documents with such things, as bibliography, footnotes, table of contents, cross references, etc. But in a same time I want to have a simple interactive Jupyter Notebook ipynb file to exchange with others etc.

This is the link to [Google Drive folder](https://drive.google.com/drive/folders/1eA4v_ZpQ-2reF0uuCcZNfu2jSycPwvpY?usp=sharing), where I upload example of ipynb file, zipped html+png output made by Quatro, and html output, made by Jupyter Lab Export.

The Quatro command is `quarto.exe render Jupyter_SymPy_Quarto_math_output.ipynb`. The format html specified in JAML section of ipynb file. The output of command is

```
pandoc 
  to: html
  output-file: Jupyter_SymPy_Quarto_math_output.html
  standalone: true
  section-divs: true
  html-math-method: mathjax
  wrap: none
  default-image-extension: png
  
metadata
  document-css: false
  link-citations: true
  date-format: long
  lang: en
  
Output created: Jupyter_SymPy_Quarto_math_output.html 
```

Here is the screenshot of Jupyter-generated html:
![jupyter-export](https://user-images.githubusercontent.com/19704851/188105908-6698d204-8772-46b6-a5d6-67591f92e8d9.png)

Here is the screenshot of Quarto-generated html:
![quarto-export](https://user-images.githubusercontent.com/19704851/188106177-680c7358-583c-467e-8c43-e01218796860.png)

I am using:
Windows 10;
Quarto 1.1.175;
Python 3.9.12, Jupyter Lab 3.3.2, sympy 1.10.1, pandoc 2.12, all installed and running with Anaconda Navigator 2.2.0.

Thanks for your attention to my problem.

**Comment by rgaiacs** [comment]
Quarto can render value of SymPy objects using LaTeX.

When you do [`init_printing()`](sympy.interactive.printing.init_printing), SymPy decide what is the best output format. From your report, SymPy doesn't know that it's being used in Quarto. You should change the output format to LaTeX by using `init_printing(use_latex='mathjax')`.

**Reply by quctex** [reply]
Thanks for your answer, rgaiacs! Now I get this output:
![pic](https://user-images.githubusercontent.com/19704851/188111633-7debb7ab-2523-40de-a0a0-53a90b19c46a.png)

All formulas now rendered as text. Maybe some small thing missing here?

**Reply by rgaiacs** [reply]
Can you provide the intermediate Markdown file?

**Reply by quctex** [reply]
Yes, I can:
[Jupyter_SymPy_Quarto_math_output.md](https://github.com/quarto-dev/quarto-cli/files/9477177/Jupyter_SymPy_Quarto_math_output.md)

**Reply by quctex** [reply]
Here I provide intermediate Markdown file in case, when PNG-formula inserted in html, as described in my start-post:
[Jupyter_SymPy_Quarto_math_output_png.md](https://github.com/quarto-dev/quarto-cli/files/9481529/Jupyter_SymPy_Quarto_math_output_png.md).

Maybe this additional information will help.

@rgaiacs, you say 

> Quarto can render value of SymPy objects **using LaTeX**.

Did you mean LaTeX-notation or LaTeX program? I have no installed any LaTeX distribution on my computer. Maybe I should install MiKTeX or TeX-Live? Although I believe that MathJax and MathML are pure web technologies, that do not need LaTeX executables. Maybe I am wrong.

**Reply by rgaiacs** [reply]
@quctex Thanks for the Markdown file.

> Did you mean LaTeX-notation or LaTeX program?

LaTeX-notation.

> Maybe I should install MiKTeX or TeX-Live?

You don't need. Quarto uses MathJax.

The answer that I provided to you was incomplete. Sorry. Based on https://groups.google.com/g/sympy/c/gVh5-yJv094/m/jAE-jZyPBAAJ, you should use something like `sp.init_printing(str_printer=sp.latex, pretty_print=False, mode='inline')`. I tested on my Python command line and it works:

```
% python
Python 3.10.6 | packaged by conda-forge | (main, Aug 22 2022, 20:36:39) [GCC 10.4.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import sympy as sp
>>> sp.__version__
1.10.1
>>> sp.init_printing(str_printer=sp.latex, pretty_print=False, mode='inline')
>>> E, J, m, omega_k = sp.symbols('E J m omega_k')
>>> (4.5*m) / (E*J) - 1/omega_k**2
$- \frac{1}{\omega_{k}^{2}} + 4.5 m / E J$
```

From the Markdown file that you provided, Jupyter will produce something like

``````
::: {.cell execution_count=4}
``` {.python .cell-code}
(4.5*m) / (E*J) - 1/omega_k**2
```

::: {.cell-output .cell-output-display execution_count=4}
```
$- \frac{1}{\omega_{k}^{2}} + 4.5 m / E J$
```
:::
:::
``````

You will need to add `output: 'asis'` to the YAML header of the code cells.

**Reply by quctex** [reply]
With `sp.init_printing(str_printer=sp.latex, pretty_print=False, mode='inline')` now I got a raw latex-code inside generated by quadro html-file:
![img](https://user-images.githubusercontent.com/19704851/188252551-fe126483-77ea-4b00-a20e-ce812d39a419.png)

Corresponding intermediate md: [Jupyter_SymPy_Quarto_math_output.md](https://github.com/quarto-dev/quarto-cli/files/9481664/Jupyter_SymPy_Quarto_math_output.md) Notice, that there is no `$` symbols around the latex-code. Also there is  no `:::` framing around the code (I do not understand the the role of these symbols).

I repeated the test that your did in pure python console, and it gave the same result, with  `$` symbols around latex code:
```
Python 3.9.12 (main, Apr  4 2022, 05:22:27) [MSC v.1916 64 bit (AMD64)] :: Anaconda, Inc. on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> import sympy as sp
>>> sp.__version__
'1.10.1'
>>> sp.init_printing(str_printer=sp.latex, pretty_print=False, mode='inline')
>>> E, J, m, omega_k = sp.symbols('E J m omega_k')
>>> (4.5*m) / (E*J) - 1/omega_k**2
$- \frac{1}{\omega_{k}^{2}} + 4.5 m / E J$
>>>
```

After that I did this test in IPython console and gave result without `$` symbols around latex code:
```
Jupyter QtConsole 5.3.0
Python 3.9.12 (main, Apr  4 2022, 05:22:27) [MSC v.1916 64 bit (AMD64)]
Type 'copyright', 'credits' or 'license' for more information
IPython 8.2.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: import sympy as sp

In [2]: sp.__version__
Out[2]: '1.10.1'

In [3]: sp.init_printing(str_printer=sp.latex, pretty_print=False, mode='inline')

In [4]: E, J, m, omega_k = sp.symbols('E J m omega_k')

In [5]: (4.5*m) / (E*J) - 1/omega_k**2
Out[5]: - \frac{1}{\omega_{k}^{2}} + \frac{4.5 m}{E J}
```

In Jupyter Lab I see the same outputs in RAW Latex code without `$` symbols:
![pic](https://user-images.githubusercontent.com/19704851/188252940-52179f9c-803e-4239-91d8-2dbb4cda59a2.png)

This is a fragment of my ipynb file:
```
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "3a73284f-7644-43f9-a5a8-02cfdfd14a25",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "- \\frac{1}{\\omega_{k}^{2}} + \\frac{4.5 m}{E J}"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#| output: asis\n",
    "(4.5*m) / (E*J) - 1/omega_k**2"
   ]
  },
```

Besides this main suspect, I noticed that:

1. what should I use `#| output: 'asis'` as in your post or `#| output: asis`, as in [documentation](https://quarto.org/docs/computations/execution-options.html#raw-output)? I have used both options in different cells with no difference.
2. According to [documentation](https://docs.sympy.org/latest/modules/interactive.html?highlight=init_printing#module-sympy.interactive.printing), `init_printing` function has no `mode` argument, but has `latex_mode={'inline'|'plain'|'equation'|'equation*'}` argument. Trying to use it as `latex_mode='plain'` with no effect. I tend to this `mode` argument refers to that last `**settings` argument, as there is no error raised with `mode` argument.

Sorry, but I don't understand your last sentence:

> From the Markdown file that you provided, Jupyter will produce something like

I am very new to Jupyter, and do not know, that Jupyter can produce something from Markdown file. Maybe Quarto produce markdown file from provided Jupyter file? In this case, my intermediate markdown file is very different as I described above.

Thank you very much for your time and attention to my problem again and again.

**Comment by rgaiacs** [comment]
From the information that you provide, IPython kernel is messing things a little bit. Need to found how to work around it.

> Also there is no `:::` framing around the code (I do not understand the the role of these symbols).

`:::` is Pandoc's way to represent HTML's `div` block.

> what should I use `#| output: 'asis'` as in your post or `#| output: asis`, as in [documentation](https://quarto.org/docs/computations/execution-options.html#raw-output)?

They are equivalent.

> According to [documentation](https://docs.sympy.org/latest/modules/interactive.html?highlight=init_printing#module-sympy.interactive.printing), `init_printing` function has no `mode` argument.

`str_printer=sp.latex` gives the function to be used for print. `sp.latex` has `mode`. You don't see a error because `mode` goes as `**`.

> From the Markdown file that you provided, Jupyter will produce something like

I was saying "based on the Markdown file ..." I didn't execute the full pipeline because I only have quarto at the computer in my office. I will try to run the full pipeline earlier next week.
